### PR TITLE
gh #67 Test profile changes for dsFPD

### PR DESCRIPTION
**Pull request by Sumip98**
Modifying L1 concerning the latest test profile changes

### CODE REVIEW

File: src/test_l1_dsFPD.c [hunk 1]
```diff
@@ -765,27 +824,33 @@ void test_l1_dsFPD_positive_dsGetFPBrightness (void)
...
 
     // Step 03: Call dsGetFPBrightness() with specific valid indicator and store the value
+    count = UT_KVP_PROFILE_GET_UINT32("dsFPD/Number_of_Indicators");
     result = dsGetFPBrightness(dsFPD_INDICATOR_POWER, &brightness_power_indicator);
     UT_ASSERT_EQUAL(result, dsERR_NONE);
 
-    // Step 04: Call dsGetFPBrightness() and loop through all valid indicators from kIndicators
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
+    // Step 04: Call dsGetFPBrightness() and loop through all valid indicators from yaml
+     for (int i = 0; i < count; i++)
```
Comment (drissya-k) [inline src/test_l1_dsFPD.c:846]: remove extra space

File: src/test_l1_dsFPD.c [hunk 2]
```diff
@@ -1033,9 +1126,11 @@ void test_l1_dsFPD_negative_dsGetFPState (void)
     UT_ASSERT_EQUAL(result, dsERR_NONE);
 
     // Step 07: Call dsGetFPState() after termination
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
+     for (int i = 0; i < count; i++)
```
Comment (drissya-k) [inline src/test_l1_dsFPD.c:1129]: extra space

File: src/test_l1_dsFPD.c [hunk 3]
```diff
@@ -338,13 +339,16 @@ void test_l1_dsFPD_negative_dsFPTerm(void)
  * |03|Call dsSetFPState() and loop through all valid indicators from kIndicators|eIndicator: [Valid Indicator], state: dsFPD_STATE_ON|dsERR_NONE| |
```
Comment (drissya-k) [inline src/test_l1_dsFPD.c:339]: Change the description of step 3

File: src/test_l1_dsFPD.c [hunk 4]
```diff
@@ -353,14 +357,17 @@ void test_l1_dsFPD_positive_dsSetFPState (void)
...
+     for (int i = 0; i < count; ++i)
+     {
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, eIndicator);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
         UT_ASSERT_EQUAL(result, dsERR_NONE);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_OFF);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_OFF);
         UT_ASSERT_EQUAL(result, dsERR_NONE);
```
Comment (drissya-k) [inline src/test_l1_dsFPD.c:371]: If FP State is "OFF", it must return dsERR_OPERATION_NOT_SUPPORTED
Same for other cases as well
https://github.com/rdkcentral/rdk-halif-device_settings/blob/3.0.0/include/dsFPD.h

File: src/test_l1_dsFPD.c [hunk 5]
```diff
@@ -393,13 +400,16 @@ void test_l1_dsFPD_positive_dsSetFPState (void)
  * |06|Terminate with dsFPTerm()| |dsERR_NONE|Ensure the system is terminated|
```
Comment (drissya-k) [inline src/test_l1_dsFPD.c:400]: Correct the description in test procedure as well

File: src/test_l1_dsFPD.c (hunk 1 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:854]: Done

File: src/test_l1_dsFPD.c (hunk 2 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:1156]: Done

File: src/test_l1_dsFPD.c (hunk 3 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

File: src/test_l1_dsFPD.c (hunk 4 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:366]: Done

File: src/test_l1_dsFPD.c (hunk 5 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:395]: Done

**Review by drissya-k — APPROVED** [review]

File: src/test_l1_dsFPD.c [hunk 6]
```diff
@@ -75,19 +75,15 @@
...
 /* Global flags to support features */
 static bool extendedEnumsSupported=false; //Default to not supported
-
-#define INDICATOR_COLORS_SIZE_1 (sizeof(kIndicatorColors_1)/sizeof(dsFPDColorConfig_t))
-#define INDICATOR_COLORS_SIZE_2 (sizeof(kIndicatorColors_2)/sizeof(dsFPDColorConfig_t))
-#define INDICATOR_COLORS_SIZE_3 (sizeof(kIndicatorColors_3)/sizeof(dsFPDColorConfig_t))
+					  //
+#define DS_FPD_KEY_SIZE 128 
 
 #define INDICATOR_COLORS_SIZE INDICATOR_COLORS_SIZE_1
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:88]: Where is this INDICATOR_COLORS_SIZE_1 define now?

File: src/test_l1_dsFPD.c [hunk 7]
```diff
@@ -107,25 +103,29 @@ static bool extendedEnumsSupported=false; //Default to not supported
...
+#define enableFPDIndicators(count) {\
     dsError_t result;\
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)\
-    {\
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);\
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);\
+    dsFPDIndicator_t eIndicator;\
+    char buffer[DS_FPD_KEY_SIZE];\
+    for (int i = 0; i < (count); i++) {\
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);\
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:111]: According to test profile files, the fp indicators start from 1, but according to the present scenario, it starts from 0, which is unavailable.  There are multiple places with similar implementation which needs to be addressed.

File: src/test_l1_dsFPD.c [hunk 8]
```diff
@@ -353,14 +352,17 @@ void test_l1_dsFPD_positive_dsSetFPState (void)
...
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
-    {
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);
+
+     // Step 03: Call dsSetFPState() and loop through all valid indicators from yaml
+     count = UT_KVP_PROFILE_GET_UINT32("dsFPD/Number_of_Indicators");
+     for (int i = 0; i < count; ++i)
+     {
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:360]: No 0 indicator type exists..as it starts from 1.

File: src/test_l1_dsFPD.c [hunk 9]
```diff
@@ -413,26 +418,32 @@ void test_l1_dsFPD_negative_dsSetFPState (void)
     result = dsSetFPState(dsFPD_INDICATOR_MAX, dsFPD_STATE_OFF);
     UT_ASSERT_EQUAL(result, dsERR_INVALID_PARAM);
     
-    // Step 04: Loop through all valid indicators from kIndicators with an invalid state value
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
+    // Retrieve the number of indicators using the macro
+    count = UT_KVP_PROFILE_GET_UINT32("dsFPD/Number_of_Indicators");
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:422]: similar indexing issue

File: src/test_l1_dsFPD.c [hunk 10]
```diff
@@ -609,39 +649,51 @@ void test_l1_dsFPD_negative_dsSetFPBlink (void)
...
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, eIndicator);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
         UT_ASSERT_EQUAL(result, dsERR_NONE);
         
-        result = dsSetFPBrightness(kIndicators[i].id, 0);
+        result = dsSetFPBrightness(eIndicator, 0);
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:685]: now brightness is part of profile file, so we can read the Max, Min values from the profile.. and add a average value (max+min)/2

MAX_BRIGHTNESS : 100            # Maximum brightness value of FPD LEDs
        MIN_BRIGHTNESS : 0              # Minimum brightness value of FPD LEDs

File: src/test_l1_dsFPD.c [hunk 11]
```diff
@@ -2836,21 +3041,23 @@ void test_l1_dsFPD_negative_dsFPSetLEDState(void)
...
     dsFPDLedState_t ledState1, ledState2;
+    int count=0;
 
     // Step 01: Initialize using dsFPInit()
     result = dsFPInit();
     UT_ASSERT_EQUAL(result, dsERR_NONE);
 
     // Step 02: Set all valid indicators to dsFPD_STATE_ON using dsSetFPState()
-    enableFPDIndicators();
+    count = UT_KVP_PROFILE_GET_UINT32("dsFPD/Number_of_Indicators");
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:3059]: As SupportedLEDStates is also defined in profile, this should be used to validate.

**Review by srinivasgtl — COMMENTED** [review]
partial review. Please fix the index change all over the file.

File: src/test_l1_dsFPD.c [hunk 12]
```diff
@@ -1058,40 +1180,60 @@ void test_l1_dsFPD_negative_dsGetFPState (void)
...
     {
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);
+	snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
         UT_ASSERT_EQUAL(result, dsERR_NONE);
 
-        dsFPDColorConfig_t *colors = kIndicators[i].supportedColors;
+        numOfSupportedColors = UT_KVP_PROFILE_GET_LIST_COUNT("dsFPD/SupportedFPDIndicators/%d/supportedColors");
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:1214]: what is %d here.  There is no indexing value or variable

File: src/test_l1_dsFPD.c [hunk 13]
```diff
@@ -1154,20 +1305,22 @@ void test_l1_dsFPD_negative_dsSetFPColor (void)
...
     {
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
         UT_ASSERT_EQUAL(result, dsERR_NONE);
 
-        dsFPDColorConfig_t *colors = kIndicators[i].supportedColors;
+        numOfSupportedColors = UT_KVP_PROFILE_GET_LIST_COUNT("dsFPD/SupportedFPDIndicators/%d/supportedColors");
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:1315]: no indexing here as well.  %d is not associated with any indexing variable or value.

File: src/test_l1_dsFPD.c (hunk 6 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

File: src/test_l1_dsFPD.c (hunk 7 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:109]: Done

File: src/test_l1_dsFPD.c (hunk 8 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:358]: Done

File: src/test_l1_dsFPD.c (hunk 9 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:420]: Done

File: src/test_l1_dsFPD.c (hunk 10 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

File: src/test_l1_dsFPD.c (hunk 11 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:3091]: Done

File: src/test_l1_dsFPD.c (hunk 12 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

File: src/test_l1_dsFPD.c (hunk 13 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

**Review by ishthiyaqahmed — COMMENTED** [review]
could you please add dsERR_OPERATION_NOT_SUPPORTED  scenario as well for all the api's where ever it is applicable.

File: src/test_l1_dsFPD.c (hunk 7 above)
Comment (ishthiyaqahmed) [reply src/test_l1_dsFPD.c:109]: sumi please check the i value since we are getting count value also 1 in that case  for loop will not execute.

Comment (Sumip98) [reply src/test_l1_dsFPD.c:109]: Done

**Review by ishthiyaqahmed — APPROVED** [review]

File: src/test_l1_dsFPD.c (hunk 11 above)
Comment (srinivasgtl) [reply src/test_l1_dsFPD.c:3088]: Please consider the SupportedLEDStates to validate the led state as well as this is now available in the test profile.

**Review by srinivasgtl — COMMENTED** [review]
please address if there are any new comments.

File: src/test_l1_dsFPD.c [hunk 14]
```diff
@@ -668,67 +731,98 @@ void test_l1_dsFPD_positive_dsSetFPBrightness (void)
...
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
+    // Variation 05: Set all valid indicators to dsFPD_STATE_OFF using dsSetFPState()
+    for(int i = 1; i <= count; i++)
     {
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_OFF);
-        UT_ASSERT_EQUAL(result, dsERR_NONE);
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_OFF);
+        UT_ASSERT_EQUAL(result, dsERR_OPERATION_NOT_SUPPORTED);
```
[the remark this answers is not in the document]
Comment (Sumip98) [reply src/test_l1_dsFPD.c]: Done

File: src/test_l1_dsFPD.c [hunk 15]
```diff
@@ -667,68 +730,99 @@ void test_l1_dsFPD_positive_dsSetFPBrightness (void)
...
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+	snprintf(minbuffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/MIN_BRIGHTNESS",eIndicator);
+        minBrightness =  UT_KVP_PROFILE_GET_UINT32(minbuffer);
+        snprintf(maxbuffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/MAX_BRIGHTNESS",eIndicator);
+        maxBrightness = UT_KVP_PROFILE_GET_UINT32(maxbuffer);
+        int avgBrightness = (maxBrightness + minBrightness) / 2;
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
+        UT_ASSERT_EQUAL(result, dsERR_NONE);
+        result = dsSetFPBrightness(eIndicator, avgBrightness);
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:807]: Why are we validating with NOT_SUPORTED here where we have kept a valid value?  Can you please look into it?  Also the value LED is not put to OFF state after use (setting the FPBrightness).

File: src/test_l1_dsFPD.c [hunk 16]
```diff
@@ -1297,34 +1501,43 @@ void test_l1_dsFPD_positive_dsGetFPColor (void)
...
-    for (int i = 0; i < sizeof(kIndicators) / sizeof(kIndicators[0]); ++i)
+    for (int i = 1; i <= count; i++)
     {
-        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, kIndicators[i].id);
-        result = dsSetFPState(kIndicators[i].id, dsFPD_STATE_ON);
+	snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+
+        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__, eIndicator);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:1538]: This was never switched off again.

File: src/test_l1_dsFPD.c [hunk 17]
```diff
@@ -1371,21 +1588,36 @@ void test_l1_dsFPD_negative_dsGetFPColor (void)
...
+    UT_ASSERT_EQUAL_FATAL(result, dsERR_NONE);
+
+    // Step 02: Set all valid indicators to dsFPD_STATE_ON using dsSetFPState()
+    count = UT_KVP_PROFILE_GET_UINT32("dsFPD/Number_of_Indicators");
+    for (int i = 1; i <= count; i++)
+    {
+        snprintf(buffer, DS_FPD_KEY_SIZE, "dsFPD/SupportedFPDIndicators/%d/Indicator_Type", i);
+        eIndicator = UT_KVP_PROFILE_GET_UINT32(buffer);
+        UT_LOG("\n In %s , Indicator: [%d]\n", __FUNCTION__,  eIndicator);
+        result = dsSetFPState(eIndicator, dsFPD_STATE_ON);
```
Comment (srinivasgtl) [inline src/test_l1_dsFPD.c:1618]: This is never switched off.

File: src/test_l1_dsFPD.c (hunk 15 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:807]: Done

File: src/test_l1_dsFPD.c (hunk 17 above)
Comment (Sumip98) [reply src/test_l1_dsFPD.c:1622]: Done

Comment (srinivasgtl) [reply src/test_l1_dsFPD.c:1622]: Please call FPD_STATE_OFF after 
  // Step 03: Call dsSetFPDMode() and loop through all valid values in dsFPDMode_t
    for (dsFPDMode_t mode = dsFPD_MODE_ANY; mode < dsFPD_MODE_MAX; ++mode)
    {
        result = dsSetFPDMode(mode);
        UT_ASSERT_EQUAL(result, dsERR_NONE);
    }

Comment (Sumip98) [reply src/test_l1_dsFPD.c:1623]: Done

**Review by srinivasgtl — APPROVED** [review]
approved

**Review by ishthiyaqahmed — APPROVED** [review]
still could see dsERR_OPERATION_NOT_SUPPORTED in the test description , which need to be changed , discussed with sumi which will be taken care in cleanup activity hence approving the PR

**Pull request merged**